File: apps/timemanager-api/src/graphql/group_palette.ts
/**
 * Shared preset palette for activity groups.
 * Keep in sync with Flutter `lib/theme/tokens/group_palette.dart`.
 */
export const GROUP_COLOR_PALETTE = [
  '#0F766E', // teal (brand)
  '#2563EB', // blue
  '#7C3AED', // violet
  '#DB2777', // pink
  '#DC2626', // red
  '#EA580C', // orange
  '#CA8A04', // yellow
  '#16A34A', // green
  '#0891B2', // cyan
  '#4B5563', // gray
] as const

export type GroupColor = (typeof GROUP_COLOR_PALETTE)[number]

const HEX_COLOR_RE = /^#[0-9A-Fa-f]{6}$/

export function isAllowedGroupColor(color: string): color is GroupColor {
  if (!HEX_COLOR_RE.test(color)) return false
  const normalized = color.toUpperCase()
  return (GROUP_COLOR_PALETTE as readonly string[]).some(
    (c) => c.toUpperCase() === normalized,
  )
}

/** Normalize to canonical `#RRGGBB` uppercase from the allowlist. */
export function normalizeGroupColor(color: string): GroupColor {
  const match = (GROUP_COLOR_PALETTE as readonly string[]).find(
    (c) => c.toUpperCase() === color.toUpperCase(),
  )
  if (!match) {
    throw new Error(`Invalid group color: ${color}`)
  }
  return match as GroupColor
}

File: libs/design_system/lib/theme/tokens/group_palette.dart
import 'package:flutter/material.dart';

/// Preset palette for named entities (groups, categories, rewards, …).
/// Keep in sync with API palette allowlists where applicable.
const List<String> kGroupColorPalette = [
  '#0F766E', // teal (brand)
  '#2563EB', // blue
  '#7C3AED', // violet
  '#DB2777', // pink
  '#DC2626', // red
  '#EA580C', // orange
  '#CA8A04', // yellow
  '#16A34A', // green
  '#0891B2', // cyan
  '#4B5563', // gray
];

/// Alias for [kGroupColorPalette] with product-neutral naming.
const List<String> kEntityColorPalette = kGroupColorPalette;

/// Parses `#RRGGBB` (or `RRGGBB`) into a [Color].
/// Falls back to the first palette color if the value is invalid.
Color parseEntityColor(String hex) {
  var value = hex.trim();
  if (value.startsWith('#')) {
    value = value.substring(1);
  }
  if (value.length == 6) {
    final parsed = int.tryParse(value, radix: 16);
    if (parsed != null) {
      return Color(0xFF000000 | parsed);
    }
  }
  return parseEntityColor(kEntityColorPalette.first);
}

/// Alias for [parseEntityColor] (historical group naming).
Color parseGroupColor(String hex) => parseEntityColor(hex);

/// Whether [hex] matches a value in the shared palette (case-insensitive).
bool isAllowedEntityColor(String hex) {
  final normalized = hex.trim().toUpperCase();
  return kEntityColorPalette.any((c) => c.toUpperCase() == normalized);
}

/// Alias for [isAllowedEntityColor] (historical group naming).
bool isAllowedGroupColor(String hex) => isAllowedEntityColor(hex);
